modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/MonadError.kt

```diff
@@ -128,20 +128,36 @@ interface MonadThrow<F> : MonadError<F, Throwable> {
    * ```
    *
    */
+  override val fx: MonadThrowFx<F>
+    get() = object : MonadThrowFx<F> {
+      override val ME: MonadThrow<F> = this@MonadThrow
+    }
+
   @Deprecated(
     "`bindingCatch` is getting renamed to `fx` for consistency with the Arrow Fx system. Use the Fx extensions for comprehensions",
-    ReplaceWith("fx")
+    ReplaceWith("fx.monadThrow")
   )
-  fun <B> bindingCatch(c: suspend MonadErrorContinuation<F, *>.() -> B): Kind<F, B> {
-    val continuation = MonadErrorContinuation<F, B>(this)
-    val wrapReturn: suspend MonadErrorContinuation<F, *>.() -> Kind<F, B> = { just(c()) }
-    wrapReturn.startCoroutine(continuation, continuation)
-    return continuation.returnedMonad()
-  }
+  fun <B> bindingCatch(c: suspend MonadThrowContinuation<F, *>.() -> B): Kind<F, B> =
+    fx.monadThrow(c)
 
+  @Deprecated(
+    "`bindingCatch` is getting renamed to `fx` for consistency with the Arrow Fx system. Use the Fx extensions for comprehensions",
+    ReplaceWith("fx.monad")
+  )
   override fun <B> binding(c: suspend MonadContinuation<F, *>.() -> B): Kind<F, B> =
-    bindingCatch { c() }
+    fx.monad(c)
 
   fun <A> Throwable.raiseNonFatal(): Kind<F, A> =
     if (NonFatal(this)) raiseError(this) else throw this
 }
+
+interface MonadThrowFx<F> : MonadFx<F> {
+  val ME: MonadThrow<F>
+  override val M: Monad<F> get() = ME
+  fun <A> monadThrow(c: suspend MonadThrowContinuation<F, *>.() -> A): Kind<F, A> {
```

Comment:
This is a great chance to replace `MonadThrowContinuation` with a new set of interfaces that don't expose internal. Same for the other continuations.

Reply:
In the draft PR you suggest doing this in a separate PR, I'd prefer to give this a shot in a follow-up PR.